Cap size of thread pool in select_algorithm to cpu count

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146071

Summary: With changes from https://github.com/pytorch/pytorch/pull/144829, we can see more autotune configs and the size of the pool can get outta hand when using the cutlass backend.

See internal discussion at: https://fburl.com/workplace/7g4vz0zy

Test Plan: `python test/inductor/test_cutlass_backend.py`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral committed:
Cap size of thread pool in select_algorithm to cpu count

Summary: With changes from https://github.com/pytorch/pytorch/pull/144829, we can see more autotune configs and the size of the pool can get outta hand when using the cutlass backend.

Test Plan: `python test/inductor/test_cutlass_backend.py`

[ghstack-poisoned]

masnesral commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1576,10 +1576,17 @@ class NoValidChoicesError(RuntimeError):
     if "TORCHINDUCTOR_COMPILE_THREADS" in os.environ:
         return int(os.environ["TORCHINDUCTOR_COMPILE_THREADS"])
-    return None
+
```
I dunno. Cap at cpu count?

Chillee commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1700,8 +1707,7 @@ def no_op(*args, **kwargs):
                 return no_op
 
-            env_workers = get_env_num_workers()
-            num_workers = env_workers if env_workers is not None else (len(choices))
```
I think it still makes sense to cap it at choices as well.

masnesral committed:
Update on "Cap size of thread pool in select_algorithm to cpu count"


Summary: With changes from https://github.com/pytorch/pytorch/pull/144829, we can see more autotune configs and the size of the pool can get outta hand when using the cutlass backend.

See internal discussion at: https://fburl.com/workplace/7g4vz0zy

Test Plan: `python test/inductor/test_cutlass_backend.py`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@masnesral I want to check to see if the cpu count is right.

I imagine in some cases we have a machine with 8 H100 cards, and each card is running the same thing. Should we use 1/8 of the CPUs in that case, unless autotune_multi_device=True?

But there are cases people use CUDA_VISIBLE_DEVICES to specify one GPU. So torch.cuda.device_count() probably isn't the best idea

masnesral commented:
> Should we use 1/8 of the CPUs in that case, unless autotune_multi_device=True?

@henrylhtsang, sorry I guess I'm not following. Aren't the CPUs used for building the kernels to benchmark? But then the actual benchmarking is serialized. I'd assume we'd want as much build parallelism as we can get, but we wouldn't expect to get much compile performance beyond cpu count, so we may as well cap at that to (hopefully) avoid overloading the box.
